### PR TITLE
tests: Add Pannotia GPU Tests

### PR DESCRIPTION
**Pull request by erin-le**
This PR adds the Pannotia GPU tests. The input graphs to the binaries were too large to be committed, so they need to be obtained separately. Instructions for how to do so are in README.md.

weekly-tests.yaml was modified to download the input files, but I am not sure if it will work.

### CODE REVIEW

**Comment by mattsinc** [comment]
Thanks for doing this, @erin-le !  Let me know when you want me to review this -- is it WIP?

**Comment by abmerop** [comment]
This looks fine to me at a quick glance.  Brad says I should not spend time on SE mode GPU stuff so I will defer mostly to @mattsinc  :) . 

Is it possible yet to do the same tests for FS mode?  I believe we have fixed a few issues with KVM on develop that would help but that should probably be discussed in #787 and #456 not here.

File: tests/gem5/gpu-pannotia-resources/pannotia-bins/color_maxmin.gem5
Comment (BobbyRBruce) [inline tests/gem5/gpu-pannotia-resources/pannotia-bins/color_maxmin.gem5:1]: You accidentally added this directory. At the very least you've got binaries here you're also `wget`ing in the script.

For the datasets, ``wget them too. Binaries and raw data doesn't play nice with git.

File: tests/gem5/gpu/README.md
Comment (BobbyRBruce) [inline tests/gem5/gpu/README.md:1]: Looking over this file: little of it makes sense (that main.py command isn't even using the gem5/gpu directory).

File: tests/gem5/gpu/README.md [hunk 1]
```diff
@@ -6,3 +6,18 @@ To run these tests by themselves, you can run the following command in the tests
...
+gsutil -m cp -r \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/README.txt" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/bc" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/color" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/floydwarshall" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/mis" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/pagerank" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/sssp" \
+.
+```
```
Comment (BobbyRBruce) [inline tests/gem5/gpu/README.md]: We shouldnt be having the user do this. Just have the python script wget the resources was needed.

Let's do this the cheap way: put the wget's (or whatever the python equivalent is) in tests_gpu_pannotia.py. Just have a little part at the top of the file that checks if tests/gem5/resources/pannotia exits, and, if not, creates the directory downloads all the binaries and datafiles into it.

File: .github/workflows/weekly-tests.yaml [hunk 2]
```diff
@@ -71,7 +71,11 @@ jobs:
                   ref: develop
             - name: Build VEGA_X86/gem5.opt
               run: scons build/VEGA_X86/gem5.opt -j`nproc`
-
+            - name: Get Pannotia inputs
+              working-directory: ${{ github.workspace }}/tests/gem5/gpu-pannotia-resources/pannotia-datasets
+              run: gsutil -m cp -r \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/README.txt" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/bc"
+                  \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/color" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/floydwarshall" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/mis"
+                  \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/pagerank" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/sssp" \ .
```
Comment (BobbyRBruce) [inline .github/workflows/weekly-tests.yaml]: `gsutil` is mostly a tool for people to interact with their Google Cloud account. I think you need to associate a Google Account. Anyway, I doubt it's installed on our runners and it's really not the tool for the job. If we were to put the download here i'd just recommend `wget`, though I have a suggestion elsewhere in this review that we should put them in "tests_gpu_pannotia.py"

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]

**Comment by mattsinc** [comment]
Thanks @BobbyRBruce for taking a pass on this.  @erin-le let me know when you'd like me to review it -- or if it's still WIP.

**Review by BobbyRBruce — APPROVED** [review]

**Comment by erin-le** [comment]
@mattsinc Sorry for the delay; it's ready for review now.

**Review by BobbyRBruce — APPROVED** [review]

File: tests/gem5/gpu/test_gpu_pannotia.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+        joinpath(binary_path, "bc.gem5"),
+        "--options",
+        joinpath(dataset_path, "1k_128k.gr"),
+    ],
+    valid_isas=(constants.vega_x86_tag,),
+    valid_hosts=(constants.host_gcn_gpu_tag,),
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Comment (mattsinc) [inline tests/gem5/gpu/test_gpu_pannotia.py:106]: Nothing wrong with this, but previously we only ran 1 input set per app in the weekly tests.  Running more is fine if we have time/resources, but technically speaking this is not necessary.  Open to discussion though.

File: tests/gem5/gpu/test_gpu_pannotia.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        joinpath(binary_path, "bc.gem5"),
+        "--options",
+        joinpath(dataset_path, "2k_1M.gr"),
+    ],
+    valid_isas=(constants.vega_x86_tag,),
+    valid_hosts=(constants.host_gcn_gpu_tag,),
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Comment (mattsinc) [inline tests/gem5/gpu/test_gpu_pannotia.py:124]: Previously for color we were running smaller inputs (https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L275) to make the tests complete in a more tractable time.  I believe these test sizes you have will likely run for multiple days.  If that is ok, then fine but just an FYI.

File: tests/gem5/gpu/test_gpu_pannotia.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+        joinpath(binary_path, "color_max.gem5"),
+        "--options",
+        f'{joinpath(dataset_path, "G3_circuit.graph")} 0',
+    ],
+    valid_isas=(constants.vega_x86_tag,),
+    valid_hosts=(constants.host_gcn_gpu_tag,),
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Comment (mattsinc) [inline tests/gem5/gpu/test_gpu_pannotia.py:196]: same for FW: https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L309

File: tests/gem5/gpu/test_gpu_pannotia.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+        joinpath(binary_path, "fw_hip.gem5"),
+        "--options",
+        f'-f {joinpath(dataset_path, "512_65536.gr")} -m default',
+    ],
+    valid_isas=(constants.vega_x86_tag,),
+    valid_hosts=(constants.host_gcn_gpu_tag,),
+    length=constants.very_long_tag,
+)
+
+# This test fails with
```
Comment (mattsinc) [inline tests/gem5/gpu/test_gpu_pannotia.py:232]: I have a fix for this locally but was not aware it was failing for people in public -- I thought it was only for the multi-chiplet support (not on mainline).

Same idea for MIS though about input size: https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L324

File: tests/gem5/gpu/test_gpu_pannotia.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        joinpath(binary_path, "pagerank.gem5"),
+        "--options",
+        f'{joinpath(dataset_path, "coAuthorsDBLP.graph")} 0',
+    ],
+    valid_isas=(constants.vega_x86_tag,),
+    valid_hosts=(constants.host_gcn_gpu_tag,),
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Comment (mattsinc) [inline tests/gem5/gpu/test_gpu_pannotia.py:309]: Same with SSSP: https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L371

**Review by mattsinc — COMMENTED** [review]
I realize this has already been merged in, but I wanted to add a few comments on things here, @erin-le .  Thanks again for all your hard work and help with this.  I think the main question is if we should run such large input sizes for Pannotia -- the test coverage is more than the smaller ones, but at very large increases in runtime (and running out of memory in some of the cases, as you have documented).

Happy to push a follow-up that makes those changes or let you do so.

**Comment by mattsinc** [comment]
@erin-le just wanted to make sure you saw these comments

I think some of what @BobbyRBruce is seeing in #1628 is related to these comments

**Comment by erin-le** [comment]
@mattsinc: The dev team has seen these comments and @Harshil2107 is currently working on making the changes

**Comment by Harshil2107** [comment]
@mattsinc: I pushed a PR #1631 with the changes you requested. I am still running the tests locally to see how long they take and if there are any errors.

**Pull request merged**